### PR TITLE
service-based creation and deletion via delete delay package

### PR DESCRIPTION
**Pull request by philbrookes**

### CODE REVIEW

File: pkg/reconciler/ingress/ingress.go [hunk 1]
```diff
@@ -411,20 +454,24 @@ func (c *Controller) desiredLeaves(ctx context.Context, root *networkingv1.Ingre
...
-			klog.Infof("Skipping service %q because it is not assigned to any cluster", service.Name)
+		serviceLocations, empty := svc.GetLocations(service)
+		if empty {
+			continue
 		}
+		serviceNames = append(serviceNames, service.Name)
+		locations = append(locations, serviceLocations...)
 
 		// Trigger reconciliation of the root ingress when this service changes.
 		c.tracker.add(root, service)
```
Comment (astefanutti) [inline pkg/reconciler/ingress/ingress.go:470]: Now that there is a Service controller, the tracker may be removed in favour of a more idiomatic way of enqueueing corresponding Ingresses when referencing Services change.

File: pkg/reconciler/service/controller.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
+package service
+
+import (
+	"context"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/listers/apps/v1"
+	v12 "k8s.io/client-go/listers/core/v1"
```
Comment (astefanutti) [inline pkg/reconciler/service/controller.go:9]: `corev1listers`?

File: pkg/reconciler/service/controller.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
+package service
+
+import (
+	"context"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/listers/apps/v1"
```
Comment (astefanutti) [inline pkg/reconciler/service/controller.go:8]: `appsv1listers`?

File: pkg/reconciler/service/controller.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+type ControllerConfig struct {
+	ServicesClient        kubernetes.ClusterInterface
+	SharedInformerFactory informers.SharedInformerFactory
+}
+
+type Controller struct {
+	queue                 workqueue.RateLimitingInterface
+	sharedInformerFactory informers.SharedInformerFactory
+	servicesClient        kubernetes.ClusterInterface
```
Comment (astefanutti) [inline pkg/reconciler/service/controller.go:55]: `coreClient`?

File: pkg/reconciler/service/service.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+	}
+
+	return nil
+}
+
+func (c *Controller) ReconcileShadowService(ctx context.Context, service *corev1.Service) error {
+	klog.Infof("reconciling shadow service: %v", service.Name)
+	modifiers.AddFinalizer(service, shadowCleanupFinalizer)
+	//find root deployment
+	deployments, err := c.servicesClient.Cluster(service.ClusterName).AppsV1().Deployments(service.Namespace).List(ctx, metav1.ListOptions{})
```
Comment (astefanutti) [inline pkg/reconciler/service/service.go:115]: The deployment lister can be used here.

File: pkg/modifiers/annotations.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package modifiers
```
Comment (astefanutti) [inline pkg/modifiers/annotations.go:1]: Could be moved in `util/metadata` or something?

File: pkg/deleteDelay/deleteDelay.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
+package deleteDelay
```
Comment (astefanutti) [inline pkg/deleteDelay/deleteDelay.go:1]: Could be moved in `util/metadata/finalizers` or something?

**Review by astefanutti — COMMENTED** [review]
Would be useful to create some e2e tests once #37 is merged.

File: pkg/reconciler/ingress/ingress.go [hunk 8]
```diff
@@ -317,20 +340,40 @@ func (c *Controller) reconcileLeaf(ctx context.Context, rootName string, ingress
...
-		removeFinalizer(ingress, cascadeCleanupFinalizer)
-	}
-
 	// Clean the current status, and then recreate if from the other leafs.
 	rootIngress = rootIf.(*networkingv1.Ingress).DeepCopy()
 	rootIngress.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{}
 	for _, o := range others {
+		klog.Infof("updating root ingress %v from leaf: %v, with deletionTimestamp: %v", rootIngress.Name, o.Name, o.DeletionTimestamp)
+		//don't include leaves that are deleting
+		if modifiers.HasFinalizer(o, deleteDelay.DeleteAtFinalizer) {
```
[the remark this answers is not in the document]
Comment (philbrookes) [reply pkg/reconciler/ingress/ingress.go:371]: when I tried it that way it proved very unreliable. Which is why I experimented with using this finalizer

File: pkg/reconciler/ingress/ingress.go (hunk 1 above)
Comment (philbrookes) [reply pkg/reconciler/ingress/ingress.go:470]: I'm not sure what you mean by this, could you elaborate please?

Comment (astefanutti) [reply pkg/reconciler/ingress/ingress.go:470]: I've had a deeper look, and I had forgotten that now we are relying on a single shared informer for the Services, so this is perfectly OK as it stands 👍🏼.

File: pkg/reconciler/service/service.go [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+
+func (c *Controller) ReconcileShadowService(ctx context.Context, service *corev1.Service) error {
+	klog.Infof("reconciling shadow service: %v", service.Name)
+	metadata.AddFinalizer(service, shadowCleanupFinalizer)
+	//find root deployment
+	deployments, err := c.deploymentLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	rootDeployment := findRootDeployment(findDeploymentsBySelector(service.Spec.Selector, deployments))
```
Comment (astefanutti) [inline pkg/reconciler/service/service.go:119]: Should Deployments be watched and the root Service enqueued when the root Deployment changed?

File: pkg/reconciler/deployment/deployment.go [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+			err = c.coreClient.Cluster(undesired.ClusterName).AppsV1().Deployments(undesired.Namespace).Delete(ctx, undesired.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+
+			key, err := cache.MetaNamespaceKeyFunc(undesired)
+			if err != nil {
+				return err
+			}
+			c.queue.Add(key)
```
Comment (astefanutti) [inline pkg/reconciler/deployment/deployment.go:105]: Is that really necessary? I would expect the informer listener to enqueue the updated Deployment?

File: pkg/reconciler/deployment/deployment.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+				}
+			}
+		}
+	}
+	return retServices, nil
+}
+
+func (c *Controller) findCurrentShadows(root *appsv1.Deployment) ([]*appsv1.Deployment, error) {
+	// Get the current deployment shadows
+	sel, err := labels.Parse(fmt.Sprintf("%s=%s", ownedByLabel, root.Name))
```
Comment (astefanutti) [inline pkg/reconciler/deployment/deployment.go:202]: `labels.SelectorFromSet` may be better here?

File: pkg/reconciler/deployment/controller.go [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+		runtime.HandleError(err)
+		return
+	}
+	for _, deployment := range deployments {
+		c.enqueue(deployment)
+	}
+}
+
+func (c *Controller) getReferencedDeployments(service *corev1.Service) ([]*appsv1.Deployment, error) {
+	deployments, err := c.deploymentLister.List(labels.Everything())
```
Comment (astefanutti) [inline pkg/reconciler/deployment/controller.go:177]: Can't `service.Spec.Selector` be used directly here?

File: pkg/reconciler/deployment/controller.go [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+
+	// Watch for events related to Deployments
+	c.sharedInformerFactory.Apps().V1().Deployments().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueue(obj) },
+	})
+
+	// Watch for events related to Services
+	c.sharedInformerFactory.Core().V1().Services().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```
Comment (astefanutti) [inline pkg/reconciler/deployment/controller.go:43]: Should `AddFunc` be added?

File: pkg/reconciler/deployment/deployment.go [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+	services, err := c.serviceLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, svc := range services {
+		if !service.IsRootService(svc) {
+			continue
+		}
+		for deploymentSelectorKey, deploymentSelectorValue := range deployment.Spec.Selector.MatchLabels {
```
Comment (astefanutti) [inline pkg/reconciler/deployment/deployment.go:189]: `Selector.Match` would be more robust?

File: pkg/reconciler/deployment/controller.go (hunk 12 above)
Comment (philbrookes) [reply pkg/reconciler/deployment/controller.go:177]: does the deploymentLister.List use the passed in selector against the matchLabels field in the deployment? I thought it scanned against labels.

File: pkg/reconciler/deployment/deployment.go (hunk 14 above)
Comment (philbrookes) [reply pkg/reconciler/deployment/deployment.go:189]: I'm not following this, sorry. Can I get a bit more detail on how I can implement that?

File: pkg/reconciler/deployment/controller.go [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+	deployments, err := c.deploymentLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	deployments = findDeploymentsBySelector(labels.SelectorFromSet(service.Spec.Selector), deployments)
+	return deployments, nil
+}
+
+func findDeploymentsBySelector(selector labels.Selector, deployments []*appsv1.Deployment) []*appsv1.Deployment {
+	retDeps := make([]*appsv1.Deployment, 0, len(deployments))
```
Comment (astefanutti) [inline pkg/reconciler/deployment/controller.go:192]: It seems easier to create an empty array, than duplicating all the Deployments, to filter them out.

**Review by astefanutti — APPROVED** [review]

**Pull request merged**